onnxruntime/core/providers/xnnpack/nn/unpool.cc

```diff
@@ -0,0 +1,266 @@
+    const auto* x_shape = x_arg.Shape();
+    if (!x_shape || x_shape->dim_size() != 4) {
+      break;
+    }
+
+    // require C, H, W to be known so we can construct the xnnpack kernel prior to Compute
+    if (!x_shape->dim(1).has_dim_value() ||
+        !x_shape->dim(2).has_dim_value() ||
+        !x_shape->dim(3).has_dim_value()) {
+      break;
+    }
+
+    ProtoHelperNodeContext nc(node);
+    OpNodeProtoHelper info(&nc);
+    PoolAttributes pool_attrs(info, "MaxUnpool", node.SinceVersion());
+
+    if (!IsPaddingTypeSupported(pool_attrs.auto_pad)) {
+      break;
+    }
+
+    supported = true;
+  } while (false);
+
+  return supported;
+}
+
+MaxUnpool::MaxUnpool(const OpKernelInfo& info)
+    : XnnpackKernel(info),
+      pool_attrs_{info, "MaxUnpool", info.node().SinceVersion()} {
+  num_inputs_ = OpKernel::Node().InputDefs().size();
+  info.GetAttrOrDefault<int64_t>("mode", &is_indice_produced_by_xnnpack_, 0);
+  // input is NHWC and we only support input with 4 dims. we checked C, H, W were all known in the op support checker
+  const auto& X_arg = *Node().InputDefs()[0];
+  auto X_shape = utils::GetTensorShapeFromTensorShapeProto(*X_arg.Shape());
+
+  int64_t H = X_shape[1];
+  int64_t W = X_shape[2];
+  int64_t C = X_shape[3];
+
+  // create NCHW shape to calculate most of the output shape. 'N' is set in Compute.
+  TensorShapeVector input_shape{1, C, H, W};
+  auto pads = pool_attrs_.pads;
+  output_dims_ = InferOutputSizeForUnPool(pool_attrs_, X_shape);
+  if (!is_indice_produced_by_xnnpack_) {
+    return;
+  }
+  if (num_inputs_ == 3) {
+    const Tensor* output_shape_tensor = nullptr;
+    ORT_ENFORCE(info.TryGetConstantInput(2, &output_shape_tensor), "Get output shape tensor failed");
+    const auto out_sp = output_shape_tensor->DataAsSpan<int64_t>();
+    if (std::accumulate(pool_attrs_.pads.begin(), pool_attrs_.pads.end(), 0LL) != 0) {
+      // use to calculate output shape for xnnpack
+      pool_attrs_.pads[0] = out_sp[2] - (H - 1) * pool_attrs_.strides[0] + pool_attrs_.kernel_shape[0];
+      pool_attrs_.pads[1] = 0;
+      pool_attrs_.pads[2] = out_sp[3] - (W - 1) * pool_attrs_.strides[1] + pool_attrs_.kernel_shape[1];
+      pool_attrs_.pads[3] = 0;
+    }
+    output_dims_.assign(out_sp.cbegin(), out_sp.cend());
+    std::swap(output_dims_[1], output_dims_[2]);
+    std::swap(output_dims_[3], output_dims_[2]);
+  }
+
+  // TEMPORARY sanity check. If C, H and W are known, the output shape should have been able to be inferred, with the
+  // exception of the batch size. Can be removed once we've run more models using xnnpack MaxUnpool.
+  auto inferred_output_shape = utils::GetTensorShapeFromTensorShapeProto(*Node().OutputDefs()[0]->Shape());
+  ORT_ENFORCE(inferred_output_shape[1] == output_dims_[1] &&
+                  inferred_output_shape[2] == output_dims_[2] &&
+                  inferred_output_shape[3] == output_dims_[3],
+              "Shape mismatch between inferred value and calculated value.");
```

Comment:
If this check is passing do we need any of the above code that is calculating the output_dims?

Reply:
Right. It's repeated checked.